src/System.Threading/src/System/Threading/ReaderWriterLock.cs

```diff
@@ -0,0 +1,1339 @@
+            {
+                Debug.Assert(lockEntry._readerLevel == 0);
+            }
+            // Check for nested reader
+            else if (lockEntry._readerLevel != 0)
+            {
+                Debug.Assert((_state & LockStates.ReadersMask) != 0);
+
+                if (lockEntry._readerLevel == MaxAcquireCount)
+                {
+                    throw new OverflowException(SR.Overflow_UInt16);
+                }
+                ++lockEntry._readerLevel;
+                return;
+            }
+            // Check if the thread already has writer lock
+            else if (_writerID == GetCurrentThreadID())
+            {
+                AcquireWriterLock(millisecondsTimeout);
+                Debug.Assert(lockEntry.IsEmpty);
+                return;
+            }
+            else
+            {
+                int spinCount = 0;
+                int currentState = _state;
+                do
+                {
+                    int knownState = currentState;
+
+                    // Reader need not wait if there are only readers and no writer
+                    if (knownState < LockStates.ReadersMask ||
+                        (
+                            (knownState & LockStates.ReaderSignaled) != 0 &&
+                            (knownState & LockStates.Writer) == 0 &&
+                            (
+                                // A waiting reader, after successfully completing the wait, expects that it can become a
+                                // reader, so ensure that there is enough room for waiting readers and this potential reader.
+                                (
+                                    (knownState & LockStates.ReadersMask) +
+                                    ((knownState & LockStates.WaitingReadersMask) >> LockStates.WaitingReadersShift)
+                                ) <= LockStates.ReadersMask - 2
+                            )
+                        ))
+                    {
+                        // Add to readers
+                        currentState = Interlocked.CompareExchange(ref _state, knownState + LockStates.Reader, knownState);
+                        if (currentState == knownState)
+                        {
+                            // One more reader
+                            break;
+                        }
+                        continue;
+                    }
+
+                    // Check for too many readers or waiting readers, or if signaling is in progress
+                    if ((knownState & LockStates.ReadersMask) == LockStates.ReadersMask ||
+                        (knownState & LockStates.WaitingReadersMask) == LockStates.WaitingReadersMask ||
+                        (knownState & LockStates.CachingEvents) == LockStates.ReaderSignaled)
+                    {
+                        // Sleep for a while, then update to the latest state and try again
+                        Helpers.Sleep(1000);
```

Comment:
Yea :). To meet one of the first two conditions, there would have to be 64K readers or waiting readers, so that should be rare enough.

The third condition is there to prevent new readers from starting to wait for a read lock while the previous set of waiting readers are being granted their lock. This I believe is necessary to guarantee thread safety for `finally` block below, under `if (!waitSucceeded)`. That block handles the case where the last waiting reader times out, to remove the ReaderSignaled bit from the state, acquire a read lock, and release it (which includes releasing other waiters as necessary).